### ISSUE TITLE
Not able to stream the output/response in chainlit

### ISSUE THREAD

**Issue by azazjhony1** [Chainlit/chainlit#500]
For some reason, I'm not able to get streaming response in chainlit UI although I do get streaming output normally in console. @willydouhard 

This is my code:

# Imports
from langchain import HuggingFacePipeline
from transformers import AutoTokenizer
import transformers
from transformers import TextStreamer
import torch
from langchain.document_loaders.csv_loader import CSVLoader
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import FAISS
from langchain.chains import ConversationalRetrievalChain
from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.prompts.prompt import PromptTemplate
import chainlit as cl
from langchain.memory import ConversationBufferMemory
from langchain.prompts import PromptTemplate, SystemMessagePromptTemplate, HumanMessagePromptTemplate,  ChatPromptTemplate
from pyngrok import ngrok
from langchain.embeddings import HuggingFaceBgeEmbeddings
from langchain.memory import ConversationBufferMemory
from langchain.llms import CTransformers
from torch import cuda, bfloat16
from langchain.llms import LlamaCpp
from langchain.callbacks.manager import CallbackManager
from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
from langchain.callbacks.manager import CallbackManager
from langchain.callbacks.streaming_stdout_final_only import FinalStreamingStdOutCallbackHandler
from langchain.chains import RetrievalQA

global memory
memory = ConversationBufferMemory(
                memory_key="chat_history",
                input_key="question",
                output_key='answer',
                return_messages=True
            )

callback_manager = CallbackManager([FinalStreamingStdOutCallbackHandler()])

# Path where the embeddings will be persisted locally
DB_FAISS_PATH = '/content/newBGE1_db_faiss'

# Installing embedding model
model_name = "BAAI/bge-small-en"
model_kwargs = {'device': 'cpu'}
encode_kwargs = {'normalize_embeddings': True} # set True to compute cosine similarity
embeddings = HuggingFaceBgeEmbeddings(
    model_name=model_name,
    model_kwargs=model_kwargs,
    encode_kwargs=encode_kwargs
)

model_id = 'meta-llama/Llama-2-7b-chat-hf'

device = f'cuda:{cuda.current_device()}' if cuda.is_available() else 'cpu'

bnb_config = transformers.BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type='nf4',
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=bfloat16
)

hf_auth = ''
model_config = transformers.AutoConfig.from_pretrained(
    model_id,
    use_auth_token=hf_auth
)

model = transformers.AutoModelForCausalLM.from_pretrained(
    model_id,
    trust_remote_code=True,
    config=model_config,
    quantization_config=bnb_config,
    device_map='auto',
    use_auth_token=hf_auth
)
model.eval()

tokenizer = transformers.AutoTokenizer.from_pretrained(
    model_id,
    use_auth_token=hf_auth,
)

streamer = TextStreamer(tokenizer, skip_prompt=True, skip_special_tokens=True)

pipeline = transformers.pipeline(
    model=model,
    tokenizer=tokenizer,
    return_full_text=True,  # langchain expects the full text
    task='text-generation',
    # we pass model parameters here too
    temperature=0.2,  # 'randomness' of outputs, 0.0 is the min and 1.0 the max
    max_new_tokens=4000,  # mex number of tokens to generate in the output
    repetition_penalty=1.1,  # without this output begins repeating
    streamer=streamer
)

  # Loading csv data
loader = CSVLoader(
    file_path="./DummyData.csv",
    csv_args={
        "delimiter": ",",
    },
  )

data = loader.load()

  # Loading pdf data
pdf_loader = PyPDFLoader("faq.pdf")
docs = pdf_loader.load()
text_splitter = RecursiveCharacterTextSplitter(chunk_size=400, chunk_overlap=15)
texts = text_splitter.split_documents(docs)

  # Embedding both these files and merging them
db = FAISS.from_documents(data, embeddings)
db2 = FAISS.from_documents(texts, embeddings)
db.merge_from(db2)

  # Saving these embeddings locally
db.save_local(DB_FAISS_PATH)

# Setting Custom Prompt

custom_template = """As a ecommerce assistant, you are instructed answer accurately and briefly to customers. If you don't know the answer just say 'I Don't know'.
Strip any special token if there, from the answer.
Just give the answers and dont say "Thank you" or anything else. Let's think step by step and accurately answer each question.

-----------------
Context:
{context}
-------------------
Chat_history:
Human: what product did john doe order?
AI:  John Doe ordered an iPhone 13 Pro Max.
Human: what is the order id of thomas?
AI:  The order ID for Thomas Black is 1008.
Human: what is the payment method of William Smith?
AI:  The payment method of William Smith is Alipay.
Human: what is the date of order of jane doe?
AI:  The date of order for Jane Doe is August 14, 2023.
Human: who ordered apple watch series 7?
AI:  The customer who ordered the Apple Watch Series 7 is William Smith.
Human: what is the order status of susan brown?
AI:  The order status of Susan Brown is Processing.
Human: what is the product description of iphone 13 pro max?
AI:  The product description of iPhone 13 Pro Max is the most powerful iPhone ever. It features a triple-lens rear camera system, a Super Retina XDR display, and a longer-lasting battery.

{chat_history}

---------------------
Answer the following question: {question}

Chatbot:"""

prompt = PromptTemplate(template=custom_template, input_variables=["context", "chat_history", "question"])

# Loading the llm model
llm = HuggingFacePipeline(pipeline = pipeline, model_kwargs = {'temperature': 0.2})

@cl.on_chat_start
async def start():
    chain = RetrievalQA.from_chain_type(llm=llm, retriever=db.as_retriever(search_type="similarity_score_threshold",
                                 search_kwargs={"score_threshold": 0.7,
                                                "k": 6}), return_source_documents=False,  chain_type="stuff",
                                  chain_type_kwargs={
        "verbose": True,
        "prompt": prompt,
        "memory": ConversationBufferMemory(
            memory_key="chat_history",
            input_key="question"),
    })
    msg = cl.Message(content="Starting the bot...")
    await msg.send()
    msg.content = "Hi, Welcome to Ecommerce Bot. What is your query?"
    await msg.update()

    cl.user_session.set("chain", chain)

@cl.on_message
async def main(message: cl.Message):
    # cb = cl.AsyncLangchainCallbackHandler(
    #     stream_final_answer=True, answer_prefix_tokens=["FINAL", "ANSWER", "Chatbot", "Assistant"]
    # )
    # cb.answer_reached = True
    # res = await chain.acall(message.content, callbacks=[cb])
    # answer = res
    # await cl.Message(content=answer).send()

    chain = cl.user_session.get("chain")
    cb = cl.LangchainCallbackHandler(
        stream_final_answer=True, answer_prefix_tokens=["FINAL", "ANSWER", "Chatbot", "Assistant", "answer"]
    )

    cb.answer_reached = True

    res = await cl.make_async(chain)(message, callbacks=[cb])

    # Do any post processing here

    await cl.Message(content=res["result"]).send()

**Comment by willydouhard** [comment]
This looks a lot like https://github.com/Chainlit/chainlit/issues/84

**Comment by azazjhony1** [comment]
hey @willydouhard, thanks for the response. I guess I'll have to find a workaround for the huggingfacepipeline since it doesn't accept callbacks.
